File: frontend/src/core/derived/class.ts
import { type Database } from 'sql.js';
import type { Operation } from '../types/operation';
import { queryAll, queryOne } from '../db/sqlite';
import type { ChangeNotification } from './index';

interface AstNode {
  type: string;
  text?: string;
  children?: AstNode[];
}

function collectText(value: unknown, parts: string[]): void {
  if (Array.isArray(value)) {
    for (const item of value) collectText(item, parts);
  } else if (value && typeof value === 'object') {
    const node = value as AstNode;
    if (node.type === 'text' && typeof node.text === 'string') {
      parts.push(node.text);
    }
    if (Array.isArray(node.children)) {
      for (const child of node.children) collectText(child, parts);
    }
  }
}

function normalizeClassName(name: unknown): string {
  if (typeof name !== 'string') return 'Untitled class';
  const trimmed = name.trim();
  if (!trimmed) return 'Untitled class';
  if (trimmed.startsWith('[')) {
    try {
      const ast = JSON.parse(trimmed) as unknown;
      if (Array.isArray(ast)) {
        const parts: string[] = [];
        collectText(ast, parts);
        return parts.join('').trim() || 'Untitled class';
      }
    } catch {
      // Not valid JSON — fall through to returning the raw string.
    }
  }
  return trimmed;
}

function computeAncestors(db: Database, classId: string, extendsClassIds: string[]): string[] {
  const ancestors = new Set<string>();
  const visited = new Set<string>([classId]);
  const stack = [...extendsClassIds];
  while (stack.length > 0) {
    const current = stack.pop() as string;
    if (visited.has(current)) continue;
    visited.add(current);
    ancestors.add(current);
    const rows = queryAll<{ ancestor_id: string }>(
      db,
      'SELECT ancestor_id FROM class_hierarchy WHERE class_id = ?',
      [current]
    );
    for (const row of rows) stack.push(row.ancestor_id);
  }
  return [...ancestors].sort();
}

function parseStoredExtends(raw: string | null): string[] {
  if (!raw) return [];
  try {
    const value = JSON.parse(raw) as unknown;
    return Array.isArray(value) ? (value as string[]) : [];
  } catch {
    return [];
  }
}

export function classExtendsWouldCycle(
  db: Database,
  classId: string,
  extendsClassIds: string[]
): string | null {
  for (const parentId of extendsClassIds) {
    if (parentId === classId) return parentId;
    const row = queryOne<{ found: number }>(
      db,
      'SELECT 1 AS found FROM class_hierarchy WHERE class_id = ? AND ancestor_id = ? LIMIT 1',
      [parentId, classId]
    );
    if (row) return parentId;
  }
  return null;
}

function applyClassHierarchy(
  db: Database,
  classId: string,
  extendsClassIds: string[],
  seen?: Set<string>
): void {
  const visitedClasses = seen ?? new Set<string>();
  if (visitedClasses.has(classId)) return;
  visitedClasses.add(classId);

  db.run('DELETE FROM class_hierarchy WHERE class_id = ?', [classId]);
  db.run(
    'INSERT OR IGNORE INTO class_hierarchy (class_id, ancestor_id) VALUES (?, ?)',
    [classId, classId]
  );
  for (const ancestorId of computeAncestors(db, classId, extendsClassIds)) {
    db.run(
      'INSERT OR IGNORE INTO class_hierarchy (class_id, ancestor_id) VALUES (?, ?)',
      [classId, ancestorId]
    );
  }

  const children = queryAll<{ id: string; extends_class_ids: string | null }>(
    db,
    'SELECT id, extends_class_ids FROM class WHERE id != ?',
    [classId]
  );
  for (const child of children) {
    const childExtends = parseStoredExtends(child.extends_class_ids);
    if (childExtends.includes(classId)) {
      applyClassHierarchy(db, child.id, childExtends, visitedClasses);
    }
  }
}

export function applyClassOperation(db: Database, op: Operation): ChangeNotification[] {
  const { opType } = op.envelope;
  const payload = op.payload as Record<string, unknown>;
  const ts = new Date().toISOString();

  if (opType === 'class.create') {
    const classId = payload.classId as string;
    const name = normalizeClassName(payload.name);
    const icon = (payload.icon as string | null | undefined) ?? null;
    const color = (payload.color as string | null | undefined) ?? null;
    const extendsClassIds = Array.isArray(payload.extends)
      ? (payload.extends as string[])
      : [];

    db.run(
      `INSERT OR REPLACE INTO class (
        id, workspace_id, name, icon, color, description, extends_class_ids,
        active, created_at, updated_at
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [classId, op.envelope.workspaceId, name, icon, color, null, JSON.stringify(extendsClassIds), 1, ts, ts]
    );
    applyClassHierarchy(db, classId, extendsClassIds);
    return [{ scope: 'class', nodeId: classId }];
  }

  if (opType === 'class.update') {
    const classId = payload.classId as string;
    const sets: string[] = [];
    const values: (string | null)[] = [];

    if ('name' in payload) {
      sets.push('name = ?');
      values.push(normalizeClassName(payload.name));
    }
    if ('icon' in payload) {
      sets.push('icon = ?');
      values.push(payload.icon as string | null);
    }
    if ('color' in payload) {
      sets.push('color = ?');
      values.push(payload.color as string | null);
    }
    if ('description' in payload) {
      sets.push('description = ?');
      values.push(payload.description as string | null);
    }

    let hierarchyChanged = false;
    if ('extends' in payload) {
      const extendsClassIds = Array.isArray(payload.extends) ? (payload.extends as string[]) : [];
      applyClassHierarchy(db, classId, extendsClassIds);
      hierarchyChanged = true;
    }

    if (sets.length === 0) {
      return hierarchyChanged ? [{ scope: 'class', nodeId: classId }] : [];
    }

    sets.push('updated_at = ?');
    values.push(ts, classId);

    db.run(`UPDATE class SET ${sets.join(', ')} WHERE id = ?`, values);
    return [{ scope: 'class', nodeId: classId }];
  }

  if (opType === 'class.delete') {
    const classId = payload.classId as string;
    db.run('UPDATE class SET active = 0, updated_at = ? WHERE id = ?', [ts, classId]);
    return [{ scope: 'class', nodeId: classId }];
  }

  if (opType === 'class.setExtends') {
    const classId = payload.classId as string;
    const extendsClassIds = Array.isArray(payload.extendsClassIds)
      ? (payload.extendsClassIds as string[])
      : Array.isArray(payload.extends)
        ? (payload.extends as string[])
        : [];

    db.run('UPDATE class SET extends_class_ids = ?, updated_at = ? WHERE id = ?', [
      JSON.stringify(extendsClassIds),
      ts,
      classId,
    ]);
    applyClassHierarchy(db, classId, extendsClassIds);
    return [{ scope: 'class', nodeId: classId, relatedIds: extendsClassIds }];
  }

  return [];
}

File: app/core/derived/class.py
# noqa: N999  # ``class`` is a reserved keyword but matches the task-specified module path.
"""Derived-state applier that materialises the dedicated ``class`` table.

Classes are now stored in their own table; ``node.kind`` is only ``page`` or
``block``. Class metadata is maintained exclusively by ``class.*`` operations.
"""

from __future__ import annotations

import json
import sqlite3
from typing import Any

from app.core.operation import Operation

from .class_hierarchy import apply_class_hierarchy


def _timestamp_from_op(op: Operation) -> str | None:
    return op.envelope.timestamp.isoformat() if op.envelope.timestamp else None


def _normalize_class_name(name: Any) -> str:
    """Return a plain-text class name.

    Some class.create payloads (especially from imports) store the name as a
    JSON-encoded AST string like ``'[{"type":"paragraph",...}]'`` instead of
    plain text. Parse that and extract the text so the class catalogue renders
    correctly and search/sort work as expected.
    """
    if not isinstance(name, str):
        return "Untitled class"
    name = name.strip()
    if not name:
        return "Untitled class"
    if name.startswith("["):
        try:
            ast = json.loads(name)
        except json.JSONDecodeError:
            return name
        if isinstance(ast, list):
            parts: list[str] = []
            _collect_text(ast, parts)
            return "".join(parts).strip() or "Untitled class"
    return name


def _collect_text(value: Any, parts: list[str]) -> None:
    """Recursively collect text from an AST document or inline node."""
    if isinstance(value, list):
        for item in value:
            _collect_text(item, parts)
    elif isinstance(value, dict):
        if value.get("type") == "text" and isinstance(value.get("text"), str):
            parts.append(value["text"])
        for children in value.get("children", []):
            _collect_text(children, parts)


def apply_class_operation(conn: sqlite3.Connection, op: Operation) -> None:
    """Maintain the ``class`` table from class operations."""
    op_type = op.envelope.op_type
    payload = op.payload
    ts = _timestamp_from_op(op)

    if op_type == "class.create":
        class_id = payload["classId"]
        name = _normalize_class_name(payload.get("name", "Untitled class"))
        icon = payload.get("icon")
        color = payload.get("color")

        conn.execute(
            """
            INSERT OR REPLACE INTO class (
                id, workspace_id, name, icon, color, description,
                extends_class_ids, active, created_at, updated_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                class_id,
                op.envelope.workspace_id,
                name,
                icon,
                color,
                None,
                "[]",
                1,
                ts,
                ts,
            ),
        )

    elif op_type == "class.update":
        class_id = payload["classId"]
        sets: list[str] = []
        values: list[Any] = []

        if "name" in payload:
            sets.append("name = ?")
            values.append(_normalize_class_name(payload["name"]))
        if "icon" in payload:
            sets.append("icon = ?")
            values.append(payload["icon"])
        if "color" in payload:
            sets.append("color = ?")
            values.append(payload["color"])
        if "description" in payload:
            sets.append("description = ?")
            values.append(payload["description"])

        if not sets:
            return

        sets.append("updated_at = ?")
        values.append(ts)
        values.append(class_id)

        conn.execute(
            f"UPDATE class SET {', '.join(sets)} WHERE id = ?",
            values,
        )

    elif op_type == "class.delete":
        class_id = payload["classId"]
        conn.execute(
            "UPDATE class SET active = 0, updated_at = ? WHERE id = ?",
            (ts, class_id),
        )

    elif op_type == "class.setExtends":
        class_id = payload["classId"]
        extends = payload.get("extendsClassIds", [])
        extends_list = extends if isinstance(extends, list) else []

        conn.execute(
            "UPDATE class SET extends_class_ids = ?, updated_at = ? WHERE id = ?",
            (json.dumps(extends_list), ts, class_id),
        )
        apply_class_hierarchy(conn, class_id, extends_list)
